### ISSUE TITLE
SyncronizationContext issues.

### ISSUE THREAD

**Issue by se5a** [hardlydifficult/Miner#26]
got some issues and some potential issues with how we're doing this currently. 
* At the moment, running in linux we're not getting any UI show up at all, this is because we're using Send() when we're already on the UI thread. 
in the base ViewModel we should be checking the context first (or maybe use post instead? not sure the pros and cons of the differences here)
```
    protected void OnPropertyChanged(
      [CallerMemberName] string propertyName = null)
    {
      SynchronizationContext requestingThreadContext = SynchronizationContext.Current;
      if (requestingThreadContext != context)
      {
        context.Send((state) =>
        {
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
        }, null);
      }
      else
      {
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
      }
    }
```

* A potential problem with how we're storing the context in the ViewModel class and setting it on initialization is that we could run into problems if a viewmodel is created by another thread. 
we could check if the CurrentContext is null here:
```
    public ViewModel()
    {
      
      context = SynchronizationContext.Current;
      if (context == null)
        throw new Exception("Viewmodel is being created by a non UI thread");
    }
```
However this means that if we *want* to create a viewmodel on the fly from another thread, then we can't. 

* Unless we're creating viewmodels before the UI has a chance to initialize (I don't think we currently are?) then we don't need to set the synch context in Program.cs:
```
SynchronizationContext.SetSynchronizationContext(new UISynchronizationContext(app));
```
UI libraries all do this when the first control is created, as long as we're not trying to initialize the viewmodels too early the above is unnecessary.

**Comment by hardlydifficult** [comment]
RE Post vs Send.  Not sure the pros and cons, we had post but switched to send as it make ETO not fail as badly. But that was before WTFB's sync updates - switching to post seems to be working okay now.

Creating ViewModels.  Good point, added code similar to your suggestion to Assert and message the correct path.

Re the last point - not sure how that relates to https://github.com/hardlydifficult/Miner/commit/721d543e70377641d50ea194b1f8dbb735e2d76b  .  But if things can be simplified, sounds good.

**Comment by hardlydifficult** [comment]
#3 still open - next step on se5a1

**Comment by se5a** [comment]
Another point to keep in mind, if we want to run unit tests without UI we may want to set the context in the test.

**Comment by se5a** [comment]
Ok I've looked at #721d543 and I've asked around and done some tests and things, 
Eto doesn't set the context itself, it lets the underlying UI framework set it. 
when running in windows this is WPF, so WPF sets the context. 
when running in linux, then it's GTK. 

I've tested removing setting the custom context in program.cs, and with Post in the Viewmodel.OnPropertyChanged. 
in Eto.Windows the Viewmodel.OnPropetyChanged throws and error +		
``` $exception	{"Exception has been thrown by the target of an invocation."}	System.Reflection.TargetInvocationException```
with an inner exception:
```+		InnerException	{"The calling thread cannot access this object because a different thread owns it."}	System.Exception {System.InvalidOperationException}```

The context at this point apears to be the UI thread" 
```{System.Threading.SynchronizationContext} System.Threading.SynchronizationContext ```

in linux it runs fine without the custom context. 
I'm kind of stumped as to why this is happening. 
we could just keep the custom context here, but I guess my OCD wants to know *why* this is needed.  ;)
I'm unsure if there's an easy way to get the owner of the object at this point. 

@wtfblub any thoughts?


Another way to do this whole thing might be to store the context in a singleton, which probably wont fix the above, but would make HD's fix to point 2 easier to accomplish (or even unnecessary)

**Comment by se5a** [comment]
Ok another issue which I've just stumbled across:
Eto will, by default update the binding on every change. 
WPF is updating only once the focus on the control has been lost or the user hits enter. 

that's not too much of a problem, except that by calling Post() every time, it causes text in textboxes to be selected after a change. 
so in a textbox, you start to type "se5a".. and it'll go through Post() after the "s" and then when it returns to the ui it'll have the "s" in the textbox selected, so the next letter typed "e" will replace the "s". 
and you can only have one letter in the textbox. 

I think that we should only be calling Post() when we need to.

**Comment by wtfblub** [comment]
Did a little bit of testing.

- On linux [eto sets the SynchronizationContext directly](https://github.com/picoe/Eto/blob/aaaa81ce6b20e602e8dd21a24fbcc4c2c66e070c/Source/Eto.Gtk/Forms/ApplicationHandler.cs#L33)
- On WinForms [eto makes sure the SynchronizationContext is set](https://github.com/picoe/Eto/blob/9e8f9163a5d470fd3ce33bcb1b2d8dbc3390e733/Source/Eto.WinForms/Forms/ApplicationHandler.cs#L127)
- On wpf the SynchronizationContext gets set after the first window is shown(OnShown event) - _Note: This is only the case when using eto. A normal wpf project has it set before_

> Unless we're creating viewmodels before the UI has a chance to initialize (I don't think we currently are?)

So ViewModels are actually created too early when using wpf with eto

> However this means that if we want to create a viewmodel on the fly from another thread, then we can't

How about using `Eto.Forms.Application.Instance.(Async)Invoke` instead of SynchronizationContext in the ViewModel?

**Comment by se5a** [comment]
Ah interesting, you found the actual code.
on linux if you look at https://github.com/picoe/Eto/blob/aaaa81ce6b20e602e8dd21a24fbcc4c2c66e070c/Source/Eto.Gtk/Forms/ApplicationHandler.cs#L32
it's checking whether it's been set or not first, then setting it if not. apparently some versions of gtk didn't include a synchronization context. 

I think setting the datacontext of the view after the view has been created in program.cs might fix the problem. rather than having the view create the viewmodel. - I'm having some issues after I merged HDs latest stuff that I'm trying to figure out before I can test this. 

using Eto.Forms.Application.Instance. in the viewmodels would require referencing eto in the viewmodel projects, which feels a little wrong to me.

**Comment by se5a** [comment]
This is what I'm trying now:
https://gist.github.com/se5a/3c03f56f75f257e3e4be61d6371d5ba5
Which I'd have expected to fix most the issues, 
but it's still throws `InvalidOperationException: The calling thread cannot access this object because a different thread owns it.`
It doesn't make a whole lot of sense to me why it's being so problematic.

**Comment by wtfblub** [comment]
The SynchronizationContext is set after the window is shown so you have to create the viewmodel in the OnShown event.

[This is the SynchronizationContext that has to be set](https://referencesource.microsoft.com/#WindowsBase/Base/System/Windows/Threading/DispatcherSynchronizationContext.cs)

**Comment by se5a** [comment]
ok yeah, overriding  OnShown in MainForm and creating the viewmodel there fixes the InvalidOperationException. 
however, when I compare the contexts they're never seem to be the same, even when I'm making changes to the ui, and it shouldnt need to Post();
```    protected void OnPropertyChanged(
      [CallerMemberName] string propertyName = null)
    {
      var callingContext = SynchronizationContext.Current;
      if(context != null )
        if(callingContext != context) //this if statement always evaluates to true. 
          context.Post((state) =>
          {
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
          }, null);
        else
          PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName)); //we never end up here. 
}```

**Comment by se5a** [comment]
Ah, if we Send not Post it fixes it: https://stackoverflow.com/a/16825774
don't fully understand that but it seems to work;.

**Comment by se5a** [comment]
Ok so synchcontext changes that I've made in my fork, which I'll push soon with some other ETO updates are:

* created the MainViewModel in Miner.App.UI.ETO.MainForm.cs by overriding OnShown():
```
protected override void OnShown(EventArgs e)
{
    base.OnShown(e);
    DataContext = new MainViewModel();
}
```
and removed `SynchronizationContext.SetSynchronizationContext(new UISynchronizationContext(app));` from each of the ETO startup Program.cs 

* changed ViewModelBase context to be static:
```
static SynchronizationContext context;

public ViewModel()
{
    if(context == null && SynchronizationContext.Current != null)
        context = SynchronizationContext.Current;
}
```
so now it'll store the SynchContext on the first time it creates any viewmodel, and we don't have any risk of having different SynchContexts getting stored in different viewmodels. 

* Changed Back to Send() instead of Post()
we shouldn't need to be asynch here, it *shouldn't* take too long to update any of the ui components that doing it synchronously is a problem .
problem with using Post() is that eto updates the binding on every change, this caused some annoying issues when trying to type in textboxes. see previous comment. 


Does anyone see any problems with this? @hardlydifficult, @wtfblub ?


Things I've learned:
* You can't compare SynchContexts
* The underlying UI frameworks all set the synchronization contexts at different times. Need to be careful when we're storing a context, that we don't try store it too soon. 
* SyncronizationContext.Send() and SynchronizationContext.Post() can have different affects depending on when the ui binding updates.

**Comment by se5a** [comment]
I guess we can close this one now, thanks for the help @wtfblub
